### PR TITLE
﻿[Windows] Enable crash reporter on MinGW builds.

### PR DESCRIPTION
**Pull request by bruvzg**
Alternative to #60944

Enables crash reporter on MinGW builds, DWARF symbols are parsed using [libbacktrace](https://github.com/ianlancetaylor/libbacktrace).

<details>
 <summary>MSVC crash log example</summary>

```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.0.alpha.custom_build (c2f836b7cf68cd38629f9f729dab717ac0a2bb0b)
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] core_bind::OS::crash (godot\core\core_bind.cpp:211)
[1] core_bind::OS::crash (godot\core\core_bind.cpp:211)
[2] MethodBindT<WebXRInterface,String const &>::ptrcall (godot\core\object\method_bind.h:341)
[3] GDScriptFunction::call (godot\modules\gdscript\gdscript_vm.cpp:630)
[4] GDScriptInstance::callp (godot\modules\gdscript\gdscript.cpp:1549)
[5] Node::_notification (godot\scene\main\node.cpp:56)
[6] Node2D::_notificationv (godot\scene\2d\node_2d.h:37)
[7] Object::notification (godot\core\object\object.cpp:849)
[8] SceneTree::_notify_group_pause (godot\scene\main\scene_tree.cpp:841)
[9] SceneTree::process (godot\scene\main\scene_tree.cpp:455)
[10] Main::iteration (godot\main\main.cpp:2716)
[11] OS_Windows::run (godot\platform\windows\os_windows.cpp:748)
[12] widechar_main (godot\platform\windows\godot_windows.cpp:175)
[13] _main (godot\platform\windows\godot_windows.cpp:199)
[14] main (godot\platform\windows\godot_windows.cpp:211)
[15] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[16] BaseThreadInitThunk
-- END OF BACKTRACE --
```

</details>

<details>
 <summary>MinGW crash log example</summary>

```
CrashHandlerException: Program crashed with signal 4
Engine version: Godot Engine v4.0.alpha.custom_build (c2f836b7cf68cd38629f9f729dab717ac0a2bb0b)
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] _gnu_exception_handler (C:/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crt_handler.c:242)
[2] core_bind::OS::crash(String const&) (core/core_bind.cpp:211)
[3] void call_with_ptr_args_helper<__UnexistingClass, String const&, 0ull>(__UnexistingClass*, void (__UnexistingClass::*)(String const&), void const**, IndexSequence<0ull>) (./core/variant/binder_common.h:257)
[4] void call_with_ptr_args<__UnexistingClass, String const&>(__UnexistingClass*, void (__UnexistingClass::*)(String const&), void const**) (./core/variant/binder_common.h:505)
[5] MethodBindT<String const&>::ptrcall(Object*, void const**, void*) (./core/object/method_bind.h:343)
[6] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (modules/gdscript/gdscript_vm.cpp:1949)
[7] GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) (modules/gdscript/gdscript.cpp:1543)
[8] bool Node::_gdvirtual__process_call<false>(double) (scene/main/node.h:235)
[9] Node::_notification(int) (scene/main/node.cpp:56)
[10] Node::_notificationv(int, bool) (./scene/main/node.h:45)
[11] CanvasItem::_notificationv(int, bool) (./scene/main/canvas_item.h:45)
[12] Node2D::_notificationv(int, bool) (scene/2d/node_2d.h:37)
[13] Object::notification(int, bool) (core/object/object.cpp:847)
[14] SceneTree::_notify_group_pause(StringName const&, int) (scene/main/scene_tree.cpp:854)
[15] SceneTree::process(double) (scene/main/scene_tree.cpp:453)
[16] Main::iteration() (main/main.cpp:2716)
[17] OS_Windows::run() (platform/windows/os_windows.cpp:748)
[18] widechar_main(int, wchar_t**) (platform/windows/godot_windows.cpp:173)
[19] _main() (platform/windows/godot_windows.cpp:197)
[20] __tmainCRTStartup (C:/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:321)
[21] WinMainCRTStartup (C:/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:176)
-- END OF BACKTRACE --
```

</details>

### CODE REVIEW

**Comment by bruvzg** [comment]
Updated to work with enabled ASLR.

**Comment by fire** [comment]
I wasn't able to test.

```
platform/windows/crash_handler_windows_signal.cpp:191:9: error: use of undeclared identifier 'SIGSEGV'
 signal(SIGSEGV, nullptr);
        ^
platform/windows/crash_handler_windows_signal.cpp:192:9: error: use of undeclared identifier 'SIGFPE'
 signal(SIGFPE, nullptr);
        ^
platform/windows/crash_handler_windows_signal.cpp:193:9: error: use of undeclared identifier 'SIGILL'
 signal(SIGILL, nullptr);
        ^
platform/windows/crash_handler_windows_signal.cpp:201:9: error: use of undeclared identifier 'SIGSEGV'
 signal(SIGSEGV, CrashHandlerException);
        ^
platform/windows/crash_handler_windows_signal.cpp:202:9: error: use of undeclared identifier 'SIGFPE'
 signal(SIGFPE, CrashHandlerException);
        ^
platform/windows/crash_handler_windows_signal.cpp:203:9: error: use of undeclared identifier 'SIGILL'
 signal(SIGILL, CrashHandlerException);
        ^
6 errors generated.
```

**Comment by bruvzg** [comment]
> I wasn't able to test.

Probably it needs `signal.h` included, seems like it depends on the GCC version. PR updated.

File: drivers/SCsub [hunk 1]
```diff
@@ -14,6 +14,8 @@ SConscript("coreaudio/SCsub")
 SConscript("pulseaudio/SCsub")
 if env["platform"] == "windows":
     SConscript("wasapi/SCsub")
+    if not env.msvc:
```
Comment (fire) [inline drivers/SCsub:17]: Would it be in scope to expand to all the platforms Godot supports? At least Linux and Mac or should it be in a different pr?

File: platform/windows/crash_handler_windows.h [hunk 2]
```diff
@@ -35,12 +35,15 @@
 #include <windows.h>
 
 // Crash handler exception only enabled with MSVC
-#if defined(DEBUG_ENABLED) && defined(MSVC)
+#if defined(DEBUG_ENABLED)
 #define CRASH_HANDLER_EXCEPTION 1
 
+#ifdef MSVC
```
Comment (akien-mga) [inline platform/windows/crash_handler_windows.h:41]: I noticed this while working on another PR - is `MSVC` actually a valid define? Usually this is `_MSC_VER`.

File: platform/windows/godot_windows.cpp [hunk 3]
```diff
@@ -210,7 +210,7 @@ int main(int argc, char **argv) {
 
 	// _argc and _argv are ignored
 	// we are going to use the WideChar version of them instead
-#ifdef CRASH_HANDLER_EXCEPTION
+#if defined(CRASH_HANDLER_EXCEPTION) && defined(MSVC)
```
Comment (akien-mga) [inline platform/windows/godot_windows.cpp:213]: Same here, shouldn't we use `_MSC_VER`?

File: platform/windows/crash_handler_windows.h (hunk 2 above)
Comment (bruvzg) [reply platform/windows/crash_handler_windows.h:41]: It's set in the `detect.py` https://github.com/godotengine/godot/blob/e413d1c9c35bec80f7244772bd63a3269f56c5a2/platform/windows/detect.py#L371, so it's valid but `_MSC_VER` is probably better.

**Comment by akien-mga** [comment]
Would we benefit from using libbacktrace on Linux too? Notably, can this help improve the Linux crash handler so we don't have it relying on `-rdynamic` to get usable backtraces (which doesn't seem needed e.g. for gdb, but is for our own crash handler).

This would need to be rebased after the `target` buildsystem changes. Not sure what option we should rely on to compile libbacktrace, `debug_symbols` maybe? That's a bit weird but `dev_build` doesn't sound much better for this.

File: platform/windows/crash_handler_windows.h (hunk 2 above)
Comment (akien-mga) [reply platform/windows/crash_handler_windows.h:41]: Ah yeah I'd remove that one from `detect.py`. We're only using it in a couple of files:
```
$ rg "ifn?def MSVC"
drivers/unix/file_access_unix.cpp
47:#ifdef MSVC

$ rg "defined\(MSVC\)"
platform/windows/crash_handler_windows.h
38:#if defined(DEBUG_ENABLED) && defined(MSVC)
```

**Comment by bruvzg** [comment]
> Would we benefit from using libbacktrace on Linux too?

Maybe, I have not tested it on Linux, but it should be supported.

File: platform/windows/crash_handler_windows.h (hunk 2 above)
Comment (bruvzg) [reply platform/windows/crash_handler_windows.h]: Rebased and replaced it with `_MSC_VER` everywhere.

**Comment by fire** [comment]
@akien-mga gentle nudge for Godot Engine version 4.2

File: platform/windows/crash_handler_windows.h [hunk 4]
```diff
@@ -35,12 +35,15 @@
 #include <windows.h>
 
 // Crash handler exception only enabled with MSVC
-#if defined(DEBUG_ENABLED) && defined(_MSC_VER)
+#if defined(DEBUG_ENABLED)
```
Comment (Calinou) [inline platform/windows/crash_handler_windows.h:38]: I noticed crash handler is disabled in release builds (also on macOS), but the Linux crash handler is available in release builds. This is needed to make https://github.com/godotengine/godot/pull/61906 effective in exported projects.

I suggest we always make the crash handler available in release builds, as release builds can be compiled with debug symbols enabled (or you can use a tool to reverse them if you have access to the debug symbols).

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/51991e20143a39e9ef0107163eaf283ca0a761ea), it works as expected using a build made with MSYS2 MinGW-w64 on Windows 11 22H2.

```
ERROR: Crash button pressed
   at: (core/core_bind.cpp:241)

================================================================
CrashHandlerException: Program crashed with signal 4
Engine version: Godot Engine v4.3.dev.custom_build (195a75d891df147c7aaa58e7319095d1322ae40c)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
Image memory base: 0x7ff6d6db0000, image file base: 0x140000000
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] _gnu_exception_handler (C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crt_handler.c:232)
[2] core_bind::OS::crash(String const&) (core/core_bind.cpp:241)
[3] void call_with_validated_variant_args_helper<__UnexistingClass, String const&, 0ull>(__UnexistingClass*, void (__UnexistingClass::*)(String const&), Variant const**, IndexSequence<0ull>) (./core/variant/binder_common.h:365)
[4] void call_with_validated_object_instance_args<__UnexistingClass, String const&>(__UnexistingClass*, void (__UnexistingClass::*)(String const&), Variant const**) (./core/variant/binder_common.h:652)
[5] MethodBindT<String const&>::validated_call(Object*, Variant const**, Variant*) const (./core/object/method_bind.h:344)
[6] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (modules/gdscript/gdscript_vm.cpp:2036)
[7] GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) (modules/gdscript/gdscript.cpp:1905)
[8] bool BaseButton::_gdvirtual__pressed_call<false>() (scene/gui/base_button.h:91)
[9] BaseButton::_pressed() (scene/gui/base_button.cpp:136)
[10] BaseButton::on_action_event(Ref<InputEvent>) (scene/gui/base_button.cpp:172)
[11] BaseButton::gui_input(Ref<InputEvent> const&) (scene/gui/base_button.cpp:69)
[12] Control::_call_gui_input(Ref<InputEvent> const&) (scene/gui/control.cpp:1797)
[13] Viewport::_gui_call_input(Control*, Ref<InputEvent> const&) (scene/main/viewport.cpp:1600)
[14] Viewport::_gui_input_event(Ref<InputEvent>) (scene/main/viewport.cpp:1869)
[15] Viewport::push_input(Ref<InputEvent> const&, bool) (scene/main/viewport.cpp:3368)
[16] Window::_window_input(Ref<InputEvent> const&) (scene/main/window.cpp:1614)
[17] void call_with_variant_args_helper<Window, Ref<InputEvent> const&, 0ull>(Window*, void (Window::*)(Ref<InputEvent> const&), Variant const**, Callable::CallError&, IndexSequence<0ull>) (./core/variant/binder_common.h:304)
[18] void call_with_variant_args<Window, Ref<InputEvent> const&>(Window*, void (Window::*)(Ref<InputEvent> const&), Variant const**, int, Callable::CallError&) (./core/variant/binder_common.h:418)
[19] CallableCustomMethodPointer<Window, Ref<InputEvent> const&>::call(Variant const**, int, Variant&, Callable::CallError&) const (./core/object/callable_method_pointer.h:98)
[20] Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const (core/variant/callable.cpp:56)
[21] Variant Callable::call<Ref<InputEvent> >(Ref<InputEvent>) const (./core/variant/variant.h:863)
[22] DisplayServerWindows::_dispatch_input_event(Ref<InputEvent> const&) (platform/windows/display_server_windows.cpp:2956)
[23] DisplayServerWindows::_dispatch_input_events(Ref<InputEvent> const&) (platform/windows/display_server_windows.cpp:2925)
[24] Input::_parse_input_event_impl(Ref<InputEvent> const&, bool) (core/input/input.cpp:759)
[25] Input::flush_buffered_events() (core/input/input.cpp:1025)
[26] DisplayServerWindows::process_events() (platform/windows/display_server_windows.cpp:2640)
[27] OS_Windows::run() (platform/windows/os_windows.cpp:1475)
[28] widechar_main(int, wchar_t**) (platform/windows/godot_windows.cpp:180)
[29] _main() (platform/windows/godot_windows.cpp:204)
[30] main (platform/windows/godot_windows.cpp:223)
[31] __tmainCRTStartup (C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:267)
[32] WinMainCRTStartup (C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:157)
-- END OF BACKTRACE --
```

File: platform/windows/crash_handler_windows_signal.cpp [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+	print_error("\n================================================================");
+	print_error(vformat("%s: Program crashed with signal %d", __FUNCTION__, signal));
+
+	// Print the engine version just before, so that people are reminded to include the version in backtrace reports.
+	if (String(VERSION_HASH).is_empty()) {
+		print_error(vformat("Engine version: %s", VERSION_FULL_NAME));
+	} else {
+		print_error(vformat("Engine version: %s (%s)", VERSION_FULL_NAME, VERSION_HASH));
+	}
+	print_error(vformat("Dumping the backtrace. %s", msg));
```
Comment (Calinou) [inline platform/windows/crash_handler_windows_signal.cpp:156]: This message is printed twice. I'd suggest removing this one so that "Dumping the backtrace" appears just above the baktrace.

```suggestion
	print_error(vformat("Dumping the backtrace. %s", msg));
```

File: platform/windows/crash_handler_windows_signal.cpp [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+	String _execpath = OS::get_singleton()->get_executable_path();
+
+	// Load process and image info to determine ASLR addresses offset.
+	MODULEINFO mi;
+	GetModuleInformation(GetCurrentProcess(), GetModuleHandle(NULL), &mi, sizeof(mi));
+	int64_t image_mem_base = reinterpret_cast<int64_t>(mi.lpBaseOfDll);
+	int64_t image_file_base = get_image_base(_execpath);
+	data.offset = image_mem_base - image_file_base;
+
+	print_error(vformat("Image memory base: 0x%x, image file base: 0x%x", image_mem_base, image_file_base));
```
Comment (Calinou) [inline platform/windows/crash_handler_windows_signal.cpp:167]: What is this information useful for? We don't have it in the other crash handlers, so I'm curious.

Comment (bruvzg) [reply platform/windows/crash_handler_windows_signal.cpp]: Values are used to correct offsets when ASLR is in use, but probably not useful for the end user. Removed this print as well.

File: platform/windows/crash_handler_windows_signal.cpp (hunk 5 above)
Comment (bruvzg) [reply platform/windows/crash_handler_windows_signal.cpp:156]: Removed.

**Review by fire — APPROVED** [review]
I approve of the concept. Did not test yet.

File: COPYRIGHT.txt [hunk 7]
```diff
@@ -293,6 +293,11 @@ Comment: jpeg-compressor
 Copyright: 2012, Rich Geldreich
 License: public-domain or Apache-2.0
 
+Files: ./thirdparty/libbacktrace/
```
Comment (AThousandShips) [inline COPYRIGHT.txt:296]: Would you mind moving this down below before `./thirdparty/libpng/` to keep them alphabetical

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
If we're adding libbacktrace, shouldn't we standardize all platforms on top of it and remove our bespoke platform-specific crash handlers? At least all platforms with DWARF debug info if I understand correctly.

**Review by akien-mga — APPROVED** [review]
